crates/sui-core/src/traffic_controller/nodefw_test_server.rs

```diff
@@ -0,0 +1,120 @@
+use axum::{
+    http::StatusCode,
+    response::IntoResponse,
+    routing::{get, post},
+    Router,
+};
+use axum::{Extension, Json};
+use chrono::{DateTime, Utc};
+use std::{collections::HashMap, net::SocketAddr, sync::Arc};
+use tokio::sync::{Mutex, Notify};
+use tokio::task::JoinHandle;
+
+#[derive(Clone)]
+struct AppState {
+    /// BlockAddress -> expiry datetime
+    blocklist: Arc<Mutex<HashMap<BlockAddress, DateTime<Utc>>>>,
+}
+
+pub struct NodeFwTestServer {
+    server_handle: Option<JoinHandle<()>>,
+    shutdown_signal: Arc<Notify>,
+    state: AppState,
+}
+
+impl NodeFwTestServer {
+    pub fn new() -> Self {
+        Self {
+            server_handle: None,
+            shutdown_signal: Arc::new(Notify::new()),
+            state: AppState {
+                blocklist: Arc::new(Mutex::new(HashMap::new())),
+            },
+        }
+    }
+
+    pub async fn start(&mut self, port: u16) {
+        let app_state = self.state.clone();
+        let app = Router::new()
+            .route("/list_addresses", get(Self::list_addresses))
+            .route("/block_addresses", post(Self::block_addresses))
+            .layer(Extension(app_state.clone()));
+
+        let shutdown_signal = self.shutdown_signal.clone();
+        let addr = SocketAddr::from(([127, 0, 0, 1], port));
+
+        let server = axum::Server::bind(&addr)
+            .serve(app.into_make_service())
+            .with_graceful_shutdown(async move {
+                shutdown_signal.notified().await;
+            });
+
+        let handle = tokio::spawn(async move {
+            if let Err(e) = server.await {
+                panic!("Server error: {}", e);
+            }
+        });
+
+        tokio::spawn(Self::periodically_remove_expired_addresses(
+            app_state.blocklist.clone(),
+        ));
+
+        self.server_handle = Some(handle);
+    }
+
+    /// Direct access api for test verification
+    pub async fn list_addresses_rpc(&self) -> Vec<BlockAddress> {
+        let blocklist = self.state.blocklist.lock().await;
+        blocklist.keys().cloned().collect()
+    }
+
+    /// Endpoint handler to list addresses
+    async fn list_addresses(Extension(state): Extension<AppState>) -> impl IntoResponse {
+        let blocklist = state.blocklist.lock().await;
+        let block_addresses = blocklist.keys().cloned().collect();
+        Json(BlockAddresses {
+            addresses: block_addresses,
+        })
+    }
+
+    async fn periodically_remove_expired_addresses(
+        blocklist: Arc<Mutex<HashMap<BlockAddress, DateTime<Utc>>>>,
+    ) {
+        loop {
+            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            let mut blocklist = blocklist.lock().await;
+            let now = Utc::now();
+            blocklist.retain(|_address, expiry| now < *expiry);
+        }
+    }
+
+    /// Endpoint handler to block addresses
+    async fn block_addresses(
+        Extension(state): Extension<AppState>,
```

Comment:
You'll likely want to prefer using [`State`](https://docs.rs/axum/latest/axum/extract/struct.State.html) for application state vs an Extension these days